### PR TITLE
feat: project pipeline display existed running link when runs repeatedly

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
project pipeline display existed running link when runs repeatedly

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJhc3NpZ25lZSI6WyIxMDAxMjA1Il19&id=322630&iterationID=1297&pId=0&type=REQUIREMENT)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |  project pipeline display existed running link when runs repeatedly            |
| 🇨🇳 中文    |   项目级流水线展示正在运行的流水线链接           |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkg/common/apis/response.go [hunk 1]
```diff
@@ -77,14 +78,17 @@ func init() {
 
 func encodeError(w http.ResponseWriter, r *http.Request, err error) {
 	var status int
```
Comment (sfwn) [inline pkg/common/apis/response.go:80]: add unit test here

File: bundle/pipeline.go [hunk 2]
```diff
@@ -61,7 +61,7 @@ func (b *Bundle) CreatePipeline(req interface{}) (*apistructs.PipelineDTO, error
 		return nil, apierrors.ErrInvoke.InternalError(err)
 	}
 	if !resp.IsOK() || !createResp.Success {
-		return nil, toAPIError(resp.StatusCode(), createResp.Error)
+		return nil, toAPIError(resp.StatusCode(), createResp.Error).SetCtx(createResp.Error.Ctx)
```
Comment (sfwn) [inline bundle/pipeline.go:64]: the args[1] of `toAPIError` is `createResp.Error`, which already contains Ctx.

File: internal/apps/dop/endpoints/pipeline.go [hunk 3]
```diff
@@ -394,26 +394,11 @@ func (e *Endpoints) pipelineRun(ctx context.Context, r *http.Request, vars map[s
...
-			return errorresp.ErrResp(err)
-		}
-
-		ctxMap, ok := apiError.Ctx().(map[string]interface{})
-		if !ok {
-			// Interface converted to map[string]interface{} fails and returns the error passed by the pipeline component
-			return errorresp.ErrResp(err)
+		runningPipelineLink, ok := e.ProjectPipelineSvc.TryGenRunningPipelineLinkFromErr(p.PipelineDTO.OrgName, p.PipelineDTO.ProjectID, p.PipelineDTO.ApplicationID, err)
+		if ok {
+			return errorresp.ErrResp(apierrors.ErrParallelRunPipeline.InvalidState(fmt.Sprintf("failed to run pipeline, there is already running: %s", runningPipelineLink)))
```
Comment (sfwn) [inline internal/apps/dop/endpoints/pipeline.go:399]: adjust your error msg, it's a little bit strange near `there is already running`

File: internal/apps/dop/endpoints/pipeline.go [hunk 4]
```diff
@@ -394,26 +394,11 @@ func (e *Endpoints) pipelineRun(ctx context.Context, r *http.Request, vars map[s
...
-		if !ok {
-			// Failed to convert to apiError type, return the error passed by the pipeline component
-			return errorresp.ErrResp(err)
-		}
-
-		ctxMap, ok := apiError.Ctx().(map[string]interface{})
-		if !ok {
-			// Interface converted to map[string]interface{} fails and returns the error passed by the pipeline component
-			return errorresp.ErrResp(err)
+		runningPipelineLink, ok := e.ProjectPipelineSvc.TryGenRunningPipelineLinkFromErr(p.PipelineDTO.OrgName, p.PipelineDTO.ProjectID, p.PipelineDTO.ApplicationID, err)
```
Comment (sfwn) [inline internal/apps/dop/endpoints/pipeline.go:397]: You write the same logic in endpoints and services.
It's better to handle it only in one place, mostly in services.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/apps/dop/endpoints/pipeline.go (hunk 3 above)
Comment (chengjoey) [reply internal/apps/dop/endpoints/pipeline.go]: change to `already running link:`
image

File: bundle/pipeline.go (hunk 2 above)
Comment (chengjoey) [reply bundle/pipeline.go]: image
adjusted `toAPIError`, add WithCtx  in `New` func

File: internal/apps/dop/endpoints/pipeline.go (hunk 4 above)
Comment (chengjoey) [reply internal/apps/dop/endpoints/pipeline.go]: adjusted, TryGenRunningPipelineLinkFromErr return `err` and `bool`

File: pkg/common/apis/response.go (hunk 1 above)
Comment (chengjoey) [reply pkg/common/apis/response.go:80]: added

**Pull request merged**